### PR TITLE
nixos/uwsm: init

### PR DESCRIPTION
**Pull request by JohnRTitor**
## Description of changes

[UWSM](https://github.com/Vladimir-csp/uwsm) is a session manager that wraps a wayland
window compositor with useful systemd units like `graphical-session-pre.target`,
`graphical-session.target`, `xdg-desktop-autostart.target`.

This is useful for Wayland Compositors that do not start
these units on these own.

Example for Hyprland:
```nix
programs.hyprland.enable = true;
programs.uwsm.enable = true;
programs.uwsm.waylandCompositors = {
  hyprland = {
    compositorPrettyName = "Hyprland";
    compositorComment = "Hyprland compositor managed by UWSM";
    compositorBinPath = "/run/current-system/sw/bin/Hyprland";
  };
};
```

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/programs/wayland/uwsm.nix
Comment (kai-tub) [inline nixos/modules/programs/wayland/uwsm.nix:63]: I personally would favor a "real" example, like Hyprland/Sway just to improve clarity.
Though I do not have a strong opinion here.

File: nixos/modules/programs/wayland/uwsm.nix
Comment (kai-tub) [inline nixos/modules/programs/wayland/uwsm.nix:73]: Here, I would maybe provide a pointer as to why we are using `/run/current-system/sw/bin` instead of `lib.getExe pkgs.Hyprland` for example.
But I don't know if this would be even more confusing or not.
I am just worried that there would be a weird corner case where somebody uses `lib.getExe pkgs.Wayfire` or similar and have weird issues because of missing plugins.

**Review by kai-tub — COMMENTED** [review]
Thank you very much for continuing to push this forward! :+1: 
In general, I think that this is a good starting point and that we can gain some experience with this solution.
We can still consider coming up with a strategy that allows `uwsm` to _see_ the packages in the `session` directory but this is something that can be done later on without changing the options. :+1:

Thanks again!

File: nixos/modules/programs/wayland/uwsm.nix
Comment (JohnRTitor) [reply nixos/modules/programs/wayland/uwsm.nix:59]: I included that just above.

https://github.com/NixOS/nixpkgs/blob/1cf2e9be457d75a71f6abfbc30dbf4b17219c5a4/nixos/modules/programs/wayland/uwsm.nix#L81-L92

File: nixos/modules/programs/wayland/uwsm.nix
Comment (JohnRTitor) [reply nixos/modules/programs/wayland/uwsm.nix:69]: Done.

**Comment by JohnRTitor** [comment]
Build NixOS manual.

image

image

image

**Comment by JohnRTitor** [comment]
> We can still consider coming up with a strategy that allows `uwsm` to _see_ the packages in the `session` directory but this is something that can be done later on without changing the options. 👍

UWSM could read from a config file in /etc, but @Vladimir-csp should make that call.

**Comment by Vladimir-csp** [comment]
> UWSM could read from a config file in /etc

What should be in this config? Extra items on top of `$XDG_DATA_DIRS`?

**Comment by JohnRTitor** [comment]
Kind of. Ideally we would like a config file to configure the default behaviour of uwsm, which could include:

- List of `variables` to always export (it is already implemented but reading from a config file would be better)

- An explicit `sessions-directory` just like how `lightdm` does (https://github.com/NixOS/nixpkgs/blob/3a2592ce1e8c3cb79695838a8d45076b8e663e56/nixos/modules/services/x11/display-managers/lightdm.nix#L47

- Shell commands to run after session is opened (optional, but would be useful)

File: nixos/modules/programs/wayland/uwsm.nix
Comment (fufexan) [inline nixos/modules/programs/wayland/uwsm.nix:43]: This should render prettier. [Source](https://github.com/NixOS/nixpkgs/blob/master/doc/README.md#admonitions)
```suggestion
      ::: {.note}
      You must configure `waylandCompositors` suboptions as well
      :::
```

File: nixos/modules/programs/wayland/uwsm.nix
Comment (fufexan) [inline nixos/modules/programs/wayland/uwsm.nix:59]: IMO we could drop the `compositor` prefix since it's already present in `waylandCompositors`.

File: nixos/modules/programs/wayland/uwsm.nix
Comment (JohnRTitor) [reply nixos/modules/programs/wayland/uwsm.nix]: Sweet.

image

**Comment by Vladimir-csp** [comment]
> Kind of. Ideally we would like a config file to configure the default behaviour of uwsm, which could include:
...

Could you describe how this scenario should work? And can it be implemented via extra variables in context of `uwsm start`?

**Comment by JohnRTitor** [comment]
I am asking for. Let's say a config file exists in `/etc/uwsm/config.json` (can be in any format to be honest, as long as it is convenient to generate using a nix expression, `ini` will also do https://noogle.dev/q?term=lib.generators.to)

```json
{
  "session-directory": "/nix/store/HASH-sessions/wayland-sessions"
  "exported-variables": "SWAYSOCK I3SOCK XCURSOR_SIZE XCURSOR_THEME"
  "extra-init": "/etc/uwsm/init-commands.sh"
  "default-session": "hyprland"
}
```

`uwsm` will export these variables to systemd/dbus environment at the end of its startup. Obviously variables like `XDG_CURRENT_DESKTOP`, `XDG_DATA_DIRS`, `PATH` should always be exported. (Related: https://github.com/Vladimir-csp/uwsm?tab=readme-ov-file#2-service-startup-notification-and-vars-set-by-compositor)

As for the session directory, uwsm will search through it first if `uwsm start Hyprland` is provided, then fallback to user environment paths.

`default-session` parameter could be used when starting something like `uwsm start default`.

**Comment by Vladimir-csp** [comment]
`uwsm-default-id` file in config hierarchy is read by uwsm for default ID.
`UWSM_FINALIZE_VARNAMES` is used to add variable names to `uwsm finalize`, and is amended by plugins.

This leaves session-directory which can be injected via custom `XDG_CONFIG_DIRS`, and extra-init: in what context would this be executed? `ExecStartPost` of `wayland-wm@.service`? If so, it can be pre-put into a drop-in.

**Comment by JohnRTitor** [comment]
> `UWSM_FINALIZE_VARNAMES` is used to add variable names to `uwsm finalize`, and is amended by plugins.

`uwsm finalize` as I understand it, the compositor is supposed to run it at the end of its startup?

So for Hyprland, a `exec-once = uwsm finalize` should work in Hyprland config. Though I'd rather not have users to modify their Hyprland config at all. Can UWSM itself not execute it? I was thinking if we could do it through this (UWSM's) NixOS option/or from the UWSM side.

In NixOS, we have this xsession-wrapper script, that is run after login into a X11 session. It does UWSM's job, but only for X11. 

https://github.com/NixOS/nixpkgs/blob/b46afe538da7262ce53b9f32b4a5c4d715d79801/nixos/modules/services/x11/display-managers/default.nix#L53-L130



> `uwsm-default-id` file in config hierarchy is read by uwsm for default ID.

So `$HOME/.config/uwsm-default-id`? In NixOS, we prefer them in `/etc`, that makes it easy to dynamically create the config file using a NixOS option.

**Comment by Vladimir-csp** [comment]
> uwsm finalize as I understand it, the compositor is supposed to run it at the end of its startup?
So for Hyprland, a exec-once = uwsm finalize should work in Hyprland config. 

Yes. It has to be executed by the compositor itself, because it relies on being in compositor's environment to get its vars. Hyprland can be configured to do everything by itself without relying on `uwsm finalize`, but you would have to tweak it's config anyway, and also disable special treatment in hyprland uwsm plugin.

Plus `uwsm finalize` also ensures correct order of downstream units, since when compositor's own autostart fires, this usually means compositor is really ready. Breakage at this point resulted in having special treatment in the first place (https://github.com/Vladimir-csp/uwsm/issues/14, https://github.com/hyprwm/Hyprland/issues/4217, https://github.com/hyprwm/Hyprland/issues/7083). Plus activation environment interaction that Hyprland does has its own issues mentioned in 7083).

> So $HOME/.config/uwsm-default-id? In NixOS, we prefer them in /etc, that makes it easy to dynamically create the config file using a NixOS option.

In the whole config hierarchy (`$XDG_CONFIG_HOME:$XDG_CONFIG_DIRS`). The most specific wins. You can put it in `/etc/xdg/`.

> In NixOS, we have this xsession-wrapper script, that is run after login into a X11 session. It does UWSM's job, but only for X11. I predict that in future, we might want UWSM (a Wayland session wrapper) to do the same...

Parts of this (profile and environment, activation environment management) is done by `wayland-wm-env@hyprland.desktop.service` at `-pre` target. Other parts (analog of running `.xsession`) should be done via user units and xdg autostart.

**Comment by JohnRTitor** [comment]
> So for Hyprland, a `exec-once = uwsm finalize` should work in Hyprland config.

@fufexan on the Hyprland side, it would really be nice if distros could configure basic config like these by putting them in `/etc`.

https://github.com/NixOS/nixpkgs/blob/c3aa7b8938b17aebd2deecf7be0636000d62a2b9/nixos/modules/programs/wayland/sway.nix#L122-L126

This would allow distros to impose their own settings, if needed.

I suppose Home Manager does have the option for generating Hyprland config in `.config/hypr` but on the NixOS side, we don't have much option.

**Comment by fufexan** [comment]
Hyprland supports reading configurations from `/etc/xdg/hypr` (and I'll soon add support for XDG_CONFIG_DIRS), so it's just a matter of adding that functionality to the module.

However, this only works if there is no configuration in `$XDG_CONFIG_HOME/hypr` or `$HOME/.config/hypr`.

**Comment by JohnRTitor** [comment]
> However, this only works if there is no configuration in `$XDG_CONFIG_HOME/hypr` or `$HOME/.config/hypr`.

Perhaps the configurations can be merged when starting the WM? I don't see anyone actually declaring their whole configuration in `/etc/xdg/hypr`, I think we can repurpose it to apply special customisation on the distro side.

**Comment by fufexan** [comment]
I'll try to implement merging configurations.

> I don't see anyone actually declaring their whole configuration in `/etc/xdg/hypr`

Not yet, since there's no option for it, so it's not very discoverable. I assume if we add the possibility there'd be at least one person who'd do it over using Home Manager. I know some people aren't too keen on HM.

**Comment by JohnRTitor** [comment]
Nice! @fufexan

As for this PR, I believe this is ready as it is, for the initial support of UWSM.

- it allows login in to a UWSM wrapped Wayland session
- the Wayland compositor needs to be configured 
- the Wayland compositor's own module (like `programs.hyprland`) must be enabled as well/or the compositor needs to be in `environment.systemPackages`.

File: nixos/modules/programs/wayland/uwsm.nix
Comment (kai-tub) [inline nixos/modules/programs/wayland/uwsm.nix:109]: Could this become an individual option?
In my opinion, it is confusing if this module changes options for other services internally without any mentions in the docs.
Especially since it would probably _still work_ for most use cases with the `dbus` implementation.
But it could still be `broker` by default if `uwsm` is enabled.
Or maybe we could define it with `lib.mkDefault` ?

**Comment by Vladimir-csp** [comment]
> Especially since it would probably still work for most use cases with the dbus implementation.

It will. The only thing is, dbus activation environment will have empty vars after graphical session ends, until user service is terminated. This may (or may not) cause issues if another graphical session is started without terminating user.

File: nixos/modules/programs/wayland/uwsm.nix
Comment (JohnRTitor) [reply nixos/modules/programs/wayland/uwsm.nix:109]: I guess I can mention it in the option description, having another option seems redundant, and people can just do 

```nix
services.dbus.implementation = lib.mkForce "dbus";
```

**Comment by JohnRTitor** [comment]
Tested like https://github.com/JohnRTitor/nix-conf/commit/c51d90fb1d5a3a37c73aef5024015b35166aa596

In `flake.nix`:
```
inputs.uwsm-test-nixpkgs.url = "github:NixOS/nixpkgs/refs/pull/335625/merge";
```

In `configuration.nix`:
```nix
  imports = [
    "${inputs.uwsm-test-nixpkgs}/nixos/modules/programs/wayland/uwsm.nix"
  ];
```

Display Manager (in my case GDM) should show an option Hyprland (UWSM) if configured properly.

File: nixos/modules/programs/wayland/uwsm.nix
Comment (kai-tub) [inline nixos/modules/programs/wayland/uwsm.nix:19]: This doesn't work for me with `sddm`.
After a few quick tests, it seems to be caused by the escaping of `"`.
But I think we can also remove `-N "${opts.prettyName}"` since `uwsm` will read the name from the `Name` entry either way (if I understand the help output correctly) and I do not see a reason why it should only be `prettyName` and not `prettyname (UWSM)`.
For the binPath, I am not sure if the `path` type already escapes the path. :thinking:

Comment (JohnRTitor) [reply nixos/modules/programs/wayland/uwsm.nix]: Try now. It did work for me on GDM though.

Comment (Vladimir-csp) [reply nixos/modules/programs/wayland/uwsm.nix]: > But I think we can also remove -N "${opts.prettyName}" since uwsm will read the name from the Name entry either way (if I understand the help output correctly) and I do not see a reason why it should only be prettyName and not prettyname (UWSM).

UWSM will not have access to desktop entry if its main argument (`"${opts.binPath}"`) is compositor's binary. In this case entry's relevant fields should be provided via `-N`, `-D`. If the main argument is another entry, then it will be parsed.

Comment (kai-tub) [reply nixos/modules/programs/wayland/uwsm.nix]: :facepalm: Yeah, true. Sorry for the confusion.
Without the `"` it does work for me, but if we want to add the pretty name, we need `"` or `'` due to the space. I will have to take a closer look at the sddm related bits to understand what is happening.

Comment (Vladimir-csp) [reply nixos/modules/programs/wayland/uwsm.nix]: Desktop entries [should use double quotes](https://specifications.freedesktop.org/desktop-entry-spec/latest/exec-variables.html)

Comment (Vladimir-csp) [reply nixos/modules/programs/wayland/uwsm.nix]: `-D` is not necessary, assuming DM will set `$XDG_CURRENT_DESKTOP` properly. And unless you want to force names with `-e`

Comment (kai-tub) [reply nixos/modules/programs/wayland/uwsm.nix]: > Desktop entries [should use double quotes](https://specifications.freedesktop.org/desktop-entry-spec/latest/exec-variables.html)

I agree. But, there seems to be a bug in the way SDDM handles it.
If I add quotes to the `Exec` entry, it stops working (currently only tested in combination with `uwsm`).
It might have something to do with this line:

https://github.com/sddm/sddm/blob/4ec29a8bba033d475f197693fac6cb0c383a1da2/data/scripts/wayland-session#L49

I first thought that simply wrapping it in `"$@"` would fix it, but that seems to cause a different issue.
I am guessing that with the quotes, it tries to interpret the entire command as a single command, and I get a 127 exit code.

I will take a closer look over the weekend.

Comment (JohnRTitor) [reply nixos/modules/programs/wayland/uwsm.nix]: I was also having an issue with GDM.

It works if I type something like in the terminal
```shell
/nix/store/zn64gjhb383imvwhh73vdzqvxzvxw01w-uwsm-0.17.4/bin/uwsm start -S -N "Hyprland" -C "Hyprland compositor managed by UWSM" -- "/run/current-system/sw/bin/Hyprland"
```

but selecting the session from GDM and entering my password makes it unable to login.

I guess there could be a character limit in GDM? CC @jtojnar 

Removing -N and -C parameters makes it work on GDM.

```shell
/nix/store/zn64gjhb383imvwhh73vdzqvxzvxw01w-uwsm-0.17.4/bin/uwsm start -S -- "/run/current-system/sw/bin/Hyprland"
```

For SDDM issue: CC @K900, KDE maintainer

Comment (Vladimir-csp) [reply nixos/modules/programs/wayland/uwsm.nix]: > I first thought that simply wrapping it in "$@" would fix it, but that seems to cause a different issue.
I am guessing that with the quotes around $@, it tries to interpret the entire command as a single command, and I get a 127 exit code.

`"$@"` is The correct way to pass argument array as is. If it is lumped in one arg, it happened before this point.

**Comment by JohnRTitor** [comment]
@kai-tub is the current state good enough? It's working for me, just waiting for a approval :)

**Review by kai-tub — APPROVED** [review]

**Pull request merged**
